### ISSUE TITLE
[RFC] FuseEP Integration for MiniMax-M3 Prefill on Ascend NPU

### ISSUE THREAD

**Issue by ClownBin** [sgl-project/sglang#39521]
## 1. Summary

This RFC proposes an opt-in Ascend FuseEP path for MiniMax-M3 W8A8 prefill in SGLang. It adds a MiniMax-M3-specific normal-mode operator, DispatchFFNCombineM3, that fuses routed-expert dispatch, the two expert FFN GEMMs, and combine into one EP operation. The existing low-latency FuseEP path remains the decode path.

The initial implementation is deliberately narrow: MiniMax-M3, ModelSlim W8A8, hidden size 6144, top-k 4, 128 routed experts, and EP/TP 16 on Ascend NPU. It does not alter the generic DeepEP path, routing algorithm, attention implementation, or model semantics.

Reference implementation:

- SGLang integration: https://github.com/ClownBin/sglang/commit/2ca0e6cc3bfcb3379926e6261d95cc6dfe9f039c
- Ascend kernel and DeepEP adapter: https://github.com/ClownBin/sgl-kernel-npu/commit/87525b7305d449dfc66272465feb856188244092
- Original experiment and launch configuration: https://github.com/ClownBin/sglang/issues/19#issuecomment-5099607728

---

## 2. Motivation

For long prefill, the conventional MoE path exposes separate dispatch, expert-GEMM, and combine phases. In the tested MiniMax-M3 deployment, those boundaries leave observable accelerator idle periods and limit NPU utilization. The data path must also accommodate MiniMax-M3's W8A8 weight layout and SwiGLU variant.

FuseEP can eliminate those boundaries by performing dispatch, FFN, and combine in one operator and HCCL EP collective. This is useful for chunked prefill, where batch work amortizes normal-mode collective setup. Decode has a different latency profile, so it continues to use FuseEP's existing low-latency fused_deep_moe mode.

The supplied 56-concurrency comparison against DeepEP reported 575.84 tok/s and 54.21 ms. Two FuseEP observations reported 645.23 tok/s / 49.88 ms and 726 tok/s / 49.20 ms. These are individual runs rather than an aggregate; the RFC therefore treats them as preliminary and requires reproducible median-based evaluation before upstream acceptance.

---

## 3. Goals

- Add an opt-in ascend_fuseep MoE A2A backend path for MiniMax-M3 prefill on Ascend NPU.
- Fuse routed-expert dispatch, W8A8 FFN execution, and combine through DispatchFFNCombineM3.
- Preserve existing router/top-k semantics, shared-expert computation, scheduler, DP-attention behavior, and decode flow.
- Handle DP-attention ranks with zero local prefill tokens without an EP collective mismatch.
- Validate operator correctness, end-to-end model outputs, invalid configurations, and performance with a documented benchmark protocol.

## 4. Non-Goals

- Replacing DeepEP or making FuseEP the default backend.
- Supporting arbitrary models, expert counts, hidden sizes, top-k values, quantization formats, or EP degrees in the first version.
- Changing MiniMax-M3 routing, sparse attention, speculative decoding, or shared-expert semantics.
- Claiming performance gains for decode, short prefill, or unqualified hardware and software versions.

---

### 5.1 Architecture

SGLang continues to compute router logits and top-k selections. When the MoE A2A backend is ascend_fuseep, FusedMoE selects the Ascend FuseEP implementation for routed experts:

~~~text
router + top-k (existing SGLang)
          |
          +--> shared experts (existing model flow)
          |
          +--> routed experts
                 |
                 +-- prefill: DispatchFFNCombineM3
                 |      dispatch -> W8A8 GEMM1/SwiGLU -> GEMM2 -> combine
                 |
                 +-- decode: existing low-latency FuseEP fused_deep_moe
          |
          +--> add routed and shared-expert outputs
~~~

DispatchFFNCombineM3 receives activations, top-k expert IDs and weights, packed W8A8 expert weights/scales, and the EP HCCL group. It performs all-to-all dispatch, two expert FFN projections with the MiniMax-M3 SwiGLU epilogue, and combine in one normal-mode invocation.

### 5.2 Operator contract and weight preparation

At model-load time, the NPU quantization path transforms and format-casts routed-expert weights and scale tensors into the layout required by FuseEP. The M3 kernel is isolated from the generic operator because it has a fixed ABI and fused activation behavior.

Initial contract:

- BF16 activations and outputs.
- INT8 W8A8 expert weights and matching scales.
- Hidden size 6144, top-k 4, 128 routed experts, and EP size 16.
- MiniMax-M3 SwiGLU behavior, including model-specific clamp and scaling rules.

Unsupported combinations must fail clearly during startup or model initialization rather than silently selecting an incompatible kernel.

### 5.3 Prefill/decode selection and DP attention

SGLANG_ENABLE_M3_FUSEEP_PREFILL=1 enables the M3 normal-mode operator only for extend/prefill batches. It requires SGLANG_NPU_FUSED_MOE_MODE=2, and server-argument validation enforces that relationship. Decode stays on the existing low-latency FuseEP path and reuses the prepared weight layout.

With DP attention, an EP rank may have zero local tokens while other ranks have work. Every EP rank must still enter the collective. An idle rank supplies a deterministic zero-valued dummy route, then discards its output. Prefill dispatch capacity uses global token counts when available so each rank provides storage sufficient for its routed-token worst case.

### 5.4 Configuration

The feature is disabled by default. A qualified deployment enables it with the existing backend flag and explicit switches:

~~~bash
export SGLANG_NPU_FUSED_MOE_MODE=2
export SGLANG_ENABLE_M3_FUSEEP_PREFILL=1
export SGLANG_DEEPEP_NUM_MAX_DISPATCH_TOKENS_PER_RANK=16
python3 -m sglang.launch_server --model-path /path/to/MiniMax-M3-w8a8 --device npu --attention-backend ascend --quantization modelslim --tp-size 16 --dp-size 4 --enable-dp-attention --enable-dp-lm-head --moe-a2a-backend ascend_fuseep --deepep-mode auto
~~~

The deployment must use a compatible Ascend CANN/toolkit and the matching FuseEP vendor operator build. Network and HCCL settings are deployment-specific and intentionally outside this RFC's public interface.

---

## 6. Alternatives considered

1. Keep DeepEP for prefill. It remains the compatible baseline, but leaves dispatch/GEMM/combine boundaries visible to the runtime.
2. Use the existing low-latency FuseEP mode for every batch. It is retained for decode, but is not the intended normal-mode long-prefill operator for this M3 layout.
3. Generalize the operator before integration. The specialized integration lets us validate lifecycle, correctness, and performance with explicit constraints; generalization should follow qualification.

---

### Correctness

- An EP16 operator test verifies routed-token counts and a deterministic reference output.
- Verify the M3 SwiGLU epilogue and W8A8 weight/scale layout against a non-fused reference.
- Compare end-to-end greedy outputs with the qualified DeepEP baseline over fixed prompts, including long prefill and DP-attention batches with idle ranks.
- Test startup failures for unsupported fused-MoE modes, layouts, and operator contracts.

### Performance

- Publish the complete server command, model revision, CANN/torch-npu/FuseEP versions, node topology, dataset, prompt/output lengths, request rate, and warm-up protocol.
- Compare DeepEP and FuseEP at several concurrency levels; report at least three repetitions and median throughput, TTFT, TPOT, and E2E latency.
- Include profiling evidence that attributes gains to reduced bubbles and operator boundaries, not a changed workload or speculative acceptance rate.
- Measure decode separately and ensure M3 prefill selection does not regress it.

Acceptance requires no correctness regression on the qualified matrix, stable collective behavior with DP attention, and reproducible improvement for the target long-prefill workload.

---

## 8. Rollout and future work

Roll out behind SGLANG_ENABLE_M3_FUSEEP_PREFILL, with DeepEP remaining the fallback. Broaden to other EP sizes, quantization schemes, or MoE architectures only after each combination has its own correctness and performance qualification.

Questions for maintainers:

1. Is a model-specific opt-in backend acceptable as the first upstream step, or should it remain in the Ascend extension until a wider operator contract is ready?
2. What CANN, torch-npu, firmware, and topology matrix should be required for the initial support statement?
3. Should the environment switches become explicit server arguments once the qualified scope expands?
